src/java.base/share/native/libjava/io_util.c

```diff
@@ -163,30 +183,31 @@ writeBytes(JNIEnv *env, jobject this, jbyteArray bytes,
 
     if (len == 0) {
         return;
-    } else if (len > BUF_SIZE) {
-        buf = malloc(len);
+    } else if (len > STACK_BUF_SIZE) {
+        buf_size = len < MAX_MALLOC_SIZE ? len : MAX_MALLOC_SIZE;
+        buf = malloc(buf_size);
         if (buf == NULL) {
             JNU_ThrowOutOfMemoryError(env, NULL);
             return;
         }
     } else {
         buf = stackBuf;
+        buf_size = STACK_BUF_SIZE;
     }
 
-    (*env)->GetByteArrayRegion(env, bytes, off, len, (jbyte *)buf);
-
-    if (!(*env)->ExceptionOccurred(env)) {
-        off = 0;
-        while (len > 0) {
+    while (len > 0) {
+        write_size = len < buf_size ? len : buf_size;
+        (*env)->GetByteArrayRegion(env, bytes, off, write_size, (jbyte*)buf);
+        if (!(*env)->ExceptionOccurred(env)) {
```

Comment:
Wouldn't this result in an infinite loop if `GetByteArrayRegion` triggered an exception and `len > 0` ?  It would never enter the `if` and `len` is never changed then...

Reply:
Good catch, you are correct. Fixed in 10f14bb3faef2b55ab59a85016874d12815f3c87.